drivers/can/stm32_can.c

```diff
@@ -0,0 +1,765 @@
+			       | (filter->rtr_mask << CAN_FIRX_STD_RTR_POS)
+			       | (1U << CAN_FIRX_STD_IDE_POS);
+		}
+	} else {
+		id = (filter->ext_id << CAN_FIRX_EXT_EXT_ID_POS)
+		     | (filter->rtr << CAN_FIRX_EXT_RTR_POS)
+		     | (1U << CAN_FIRX_EXT_IDE_POS);
+
+		if (filter->ext_id_mask == CAN_EXT_ID_MASK && filter->rtr_mask) {
+			filter_type = CAN_FILTER_EXTENDED;
+		} else {
+			filter_type = CAN_FILTER_EXTENDED_MASKED;
+			mask = (filter->ext_id_mask << CAN_FIRX_EXT_EXT_ID_POS)
+			       | (filter->rtr_mask << CAN_FIRX_EXT_RTR_POS)
+			       | (1U << CAN_FIRX_EXT_IDE_POS);
+		}
+	}
+	SYS_LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->ext_id,
+		    filter->ext_id_mask);
+	SYS_LOG_DBG("Filter type: %s ID %s mask (%d)",
+		    (filter_type == CAN_FILTER_STANDARD ||
+		     filter_type == CAN_FILTER_STANDARD_MASKED) ?
+		    "standard" : "extended",
+		    (filter_type == CAN_FILTER_STANDARD_MASKED ||
+		     filter_type == CAN_FILTER_EXTENDED_MASKED) ?
+		    "with" : "without",
+		    filter_type);
+	do {
+		u64_t usage;
+
+		bank_nr = filter_nr / 4;
+		usage = (device_data->filter_usage >> filter_nr);
+		if (filter_type == CAN_FILTER_STANDARD) {
+			if (usage & 0x01) {
+				if (CAN_BANK_IS_EMPTY(device_data->filter_usage, bank_nr) ||
+				    CAN_IN_16BIT_LIST_MODE(can, bank_nr)) {
+					device_data->filter_usage &= ~(0x01ULL << filter_nr);
+					break;
+				} else {
+					/* Filter Bank has unsuitable configuration -> next bank */
+					filter_nr = (bank_nr + 1) * 4;
+				}
+			} else {
+				filter_nr += 1;
+			}
+		} else if (filter_type == CAN_FILTER_STANDARD_MASKED) {
+			if ((usage & 0x03) == 0x03) {
+				if (CAN_BANK_IS_EMPTY(device_data->filter_usage, bank_nr) ||
+				    CAN_IN_16BIT_MASK_MODE(can, bank_nr)) {
+					device_data->filter_usage &= ~(0x03ULL << filter_nr);
+					break;
+				} else {
+					filter_nr = (bank_nr + 1) * 4;
+				}
+			} else {
+				filter_nr += 2;
+			}
+		} else if (filter_type == CAN_FILTER_EXTENDED) {
+			if ((usage & 0x03) == 0x03) {
+				if (CAN_BANK_IS_EMPTY(device_data->filter_usage, bank_nr) ||
+				    CAN_IN_32BIT_LIST_MODE(can, bank_nr)) {
+					device_data->filter_usage &= ~(0x03ULL << filter_nr);
+					break;
+				} else {
+					filter_nr = (bank_nr + 1) * 4;
+				}
+			} else {
+				filter_nr += 2;
+			}
+		} else if (filter_type == CAN_FILTER_EXTENDED_MASKED) {
+			if ((usage & 0x0F) == 0x0F) {
+				device_data->filter_usage &= ~(0x0FULL << filter_nr);
+				break;
+			}
+			filter_nr += 4;
+		}
+		if (!usage) {
+			SYS_LOG_INF("No free filter bank found");
+			return CAN_NO_FREE_FILTER;
+		}
+	} while (filter_nr < CAN_MAX_NUMBER_OF_FILTES);
+
+	/* set the filter init mode */
+	bank_bit = (1U << bank_nr);
+	can->FMR |= CAN_FMR_FINIT;
+	can->FA1R &= ~bank_bit;
+
+	enum can_filter_type old_filter_type = can_get_filter_type(bank_bit,
+														  can->FM1R, can->FS1R);
+	/* TODO fifo balancing */
+	if (filter_type != old_filter_type) {
```

Comment:
you could rewrite this code block to gain in clarity as well. Externalizing the logic into a dedicated inline function.

Reply:
Agree!